vmas/simulator/dynamics/kinematic_bicycle.py

```diff
@@ -34,69 +34,74 @@ def __init__(
         return 2
 
     def process_action(self):
         # Extracts the velocity and steering angle from the agent's actions and convert them to physical force and torque
-        velocity = self.agent.action.u[:, 0]
-        steering_angle = self.agent.action.u[:, 1]
-
+        v_command = self.agent.action.u[:, 0]
+        # Ensure speed is within bounds
+        v_command = torch.clamp(
+            v_command, -self.agent.max_speed, self.agent.max_speed
+        )
+        steering_command =  self.agent.action.u[:, 1]
         # Ensure steering angle is within bounds
-        steering_angle = torch.clamp(
-            steering_angle, -self.max_steering_angle, self.max_steering_angle
+        steering_command = torch.clamp(
+            steering_command, -self.max_steering_angle, self.max_steering_angle
         )
 
         # Current state of the agent
         state = torch.cat((self.agent.state.pos, self.agent.state.rot), dim=1)
 
-        def f(state):
-            theta = state[:, 2]  # jaw angle
-            beta = torch.atan(
-                torch.tan(steering_angle) * self.l_r / (self.l_f + self.l_r)
-            )  # slip angle
-            dx = velocity * torch.cos(theta + beta)
-            dy = velocity * torch.sin(theta + beta)
-            dtheta = velocity / (self.l_f + self.l_r) * torch.sin(beta)
-            return torch.stack(
-                (dx, dy, dtheta), dim=-1
-            )  # Should return torch.Size([batch_size,3])
-
-        # Select the integration method
+        v_cur_x = self.agent.state.vel[:, 0] # Current velocity in x-direction
+        v_cur_y = self.agent.state.vel[:, 1] # Current velocity in y-direction
+        v_cur_angular = self.agent.state.ang_vel[:, 0] # Current angular velocity
+
+        # Select the integration method to calculate the change in state
         if self.integration == "euler":
-            new_state = self.euler(f, state)
+            state_change = self.euler(state, steering_command, v_command)
         else:
-            new_state = self.runge_kutta(f, state)
-
-        # Calculate the change in state
-        delta_state = new_state - state
+            state_change = self.runge_kutta(state, steering_command, v_command)
 
-        # Calculate the accelerations required to achieve the change in state
-        acceleration_x = delta_state[:, 0] / self.dt
-        acceleration_y = delta_state[:, 1] / self.dt
-        angular_acceleration = delta_state[:, 2] / self.dt
+        # Calculate the accelerations required to achieve the change in state.
+        # Note that in `core.py`, position is updated by `pos += vel_new * dt` with `vel_new` being calculated by `vel_new += force / mass`. The same principle applies to rotation.
+        acceleration_x = (state_change[:, 0] - v_cur_x * self.dt) / self.dt**2
+        acceleration_y = (state_change[:, 1] - v_cur_y * self.dt) / self.dt**2
+        acceleration_angular = (state_change[:, 2] - v_cur_angular * self.dt) / self.dt**2
```

Comment:
thanks so much for this, I need to fix it in `diff_drive` too

Reply:
You're welcome!